### ISSUE TITLE
[css-animations-2, css-transitions-2] Entry and exit animations for top-layer elements

### ISSUE THREAD

**Issue by chrishtr** [w3c/csswg-drafts#8189]
Animating an element into and out of the top layer provides context to the user for a more useful and pleasant interaction. Currently, developers have no way to do so in a way that preserves top-layer status during the animation, and as a result, demos like [this one](https://output.jsbin.com/veyoruc/quiet) won’t work in cases where exiting the top layer causes a visible change of positioning or z-index. Let's fix that.

# TL;DR of proposal

Introduce a new CSS property called `top-layer` that can be targeted by web developers in their CSS transitions and animations.

This CSS property is *not* generally available to web developers and will always be set with an `!important` UA style sheet rule. It’s defined as a CSS property for the purpose of a layered definition that fits within existing animations APIs and concepts, and as a result is easy for developers to use with existing animation APIs.

Example: a `<dialog>` remaining in the top layer during a 200ms opacity animation during close. The part marked as "New!" is the only addition needed for the developer to delay top-layer removal until the end of the animation.

```
@keyframes open {
  from { opacity: 0; }
  /* Ending at opacity: 1 is implicit, per existing CSS animations spec */
}
dialog:closed {
  opacity: 1; /* UA default */
  animation: open 200ms;
  transition: top-layer 200ms; /* New! */
}
@keyframes close {
  from { display: block; }
 /* Starting from opacity 1 is implicit, per existing CSS animations spec */ 
  to { opacity: 0; }
}
dialog {
  animation: close 200ms;
}
```

(full demo except for `top-layer` [here](https://output.jsbin.com/veyoruc/quiet))

# Details

## Use case

This feature enables support for animation when entering or leaving the top layer. Such animations often wish to preserve top-layer rendering during the course of an “exit”  or "entry" animation. Without this feature, it is impossible to perform such an animation.

Top-layer APIs: `<dialog>`, fullscreen, and [popover](https://github.com/whatwg/html/issues/7785).

Note: The demo in the TL;DR uses CSS animations, but a CSS transition would additionally require [this issue](https://github.com/w3c/csswg-drafts/issues/8174) to be resolved to be usable for entry transition animations. We also need to support animating `display:none` for top layer animations that need it, such as `<dialog>` (tracked [here](https://github.com/w3c/csswg-drafts/issues/6429)).

## Background on the top layer

The top layer is defined [here](https://fullscreen.spec.whatwg.org/#top-layer). It is currently accessed only via the Fullscreen API ([here](https://fullscreen.spec.whatwg.org/#ref-for-fullscreen-an-element)) and the `<dialog>` element’s `showModal` method ([here](https://html.spec.whatwg.org/multipage/interactive-elements.html#the-dialog-element)). When opening a fullscreen element, the fullscreen spec says “add it to its node document’s top layer”. When closing fullscreen, it says “remove it from its node document’s top layer”.

When calling `showModal`, the `<dialog>` spec says “add subject [the dialog element] to subject's node document's top layer”. When closing a dialog, it says “If subject is in its Document's top layer, then remove it”. If an element is already in the top layer, re-adding it to the top layer [moves it to the top of the top layer](https://fullscreen.spec.whatwg.org/#top-layer-add).

## Proposed Definition of the `top-layer` CSS property

The `top-layer` CSS property determines whether an element is in the top layer. It has two values:

```
top-layer: browser; /* element is in the top layer */
top-layer: none; /* element is not in the top layer */
```

* Default: `none`
* Animatable: yes
* Discrete: yes
* Accessible to developers: only via `transition`  CSS property values

When an element is “added to the top layer” (see previous section), it is placed in the ordered set of top-layer elements, but the *rendering* effect of the top layer (putting it in the top-layer stacking context and obeying the rendering order of the ordered set) only applies when top-layer is set.

The element is only removed from the ordered set once top-layer’s computed style has evaluated to `none` (at the end of the transition, if any). However, the “moved to the top of the top layer” behavior still occurs if the element is “added to the top layer” while animating out (see previous section).

The following UA style rules are added:

```
dialog {
 top-layer: none !important;
}

dialog:modal:open {
  top-layer: browser !important;
}

:fullscreen {
  top-layer: browser !important;
}
```

# Why shouldn't developers be able to change `top-layer` outside of transitions?

 It’s important to restrict direct developer access because the top layer is a UA-managed resource. Not giving developers direct access guarantees that the UA is able to show top layer content on top of other content, and control eviction from the top layer when needed. This point has been discussed in #6965. If in the future a "developer" top layer is added below the browser one, we could potentially add a third value to `top-layer`.

**Comment by andruud** [comment]
I thought the concept of "UA-private except for transitions" seemed odd at first, but I suppose it could be expressed as adding a UA rule (before the other rules):

```
* {
  top-layer: initial !important;
}
```

**Comment by chrishtr** [comment]
> Good idea! I think this should work, because transitions are at the [highest priority of the cascade](https://drafts.csswg.org/css-cascade/#cascade-origin). It would also avoid introducing a new "UA private" concept, which is a plus.

I went ahead and edited the original proposal to remove "UA private" and replace with `!important` UA style sheet rules. I tested in Chromium and found that transition animations work as expected.

**Comment by tabatkins** [comment]
...huh. This seems like an incredibly clever hack. I'm not opposed to it!

**Comment by jakearchibald** [comment]
I think this complicates when a dialog is 'closed', since it creates a temporal weird zone where the dialog is closed but still 'there'. Fwiw, View Transitions was designed to allow transitions for things that are no longer there, so I'm not sure why we're trying to create another thing for the same use-case.

**Comment by khushalsagar** [comment]
> When an element is “added to the top layer” (see previous section), it is placed in the ordered set of top-layer elements, but the rendering effect of the top layer (putting it in the top-layer stacking context and obeying the rendering order of the ordered set) only applies when top-layer is set.

I'm unclear on the exact effect of adding this element to the the ordered set but not changing its rendering order. There is a set of rules defined [here](https://fullscreen.spec.whatwg.org/#new-stacking-layer:~:text=Each%20element%20and%20%3A%3Abackdrop%20pseudo%2Delement%20in%20a%20top%20layer%20has%20the%20following%20characteristics%3A), the change in rendering order (by updating the element's parent stacking context) is part of it. But isn't all of this deferred by the `transition` property which delays updating the `top-layer` value?

**Comment by LeaVerou** [comment]
> > I thought the concept of "UA-private except for transitions" seemed odd at first, but I suppose it could be expressed as adding a UA rule (before the other rules):
> 
> Good idea! I think this should work, because transitions are at the [highest priority of the cascade](https://drafts.csswg.org/css-cascade/#cascade-origin). It would also avoid introducing a new "UA private" concept, which is a plus.

What prevents developers from applying an animation like:

```css
@keyframes foo {
	from, to { top-layer: browser }
}

.foo {
	animation: foo 1s infinite paused both; /* actually any of these keywords by itself works too */
}
```

to put elements in the top-layer?

**Comment by emilio** [comment]
@LeaVerou that'd be in the animations, not transition origin.

**Comment by LeaVerou** [comment]
@emilio Then `transition: 999999999s top-layer` to make it last forever after a given trigger 🤷🏽‍♀️

**Comment by nt1m** [comment]
@chrishtr mentioned that was objection around appending things to top layer during the transition. I believe the objection was more around the `:top-layer` pseudo and the interactions around it regarding transition (e.g. `:top-layer` would be matching even when the popup isn't open). Now that the pseudo-class is named `:open`/`:closed`, I think this approach can potentially be revisited.

cc @fantasai who brought this up during the call.

**Comment by nt1m** [comment]
Also, I'd like to re-iterate my objection to exposing the name "top layer" explicitly to the web platform.

These are the reasons why:

* There is general confusion with the name "top layer", e.g. is it the "topmost" element in the top-layer, or is this every element in the top layer? (In the specs, it refers to the the whole stack of elements).
* Top layer was meant as an UA implementation detail to alleviate `z-index` issues. It's OK to document it, just like we document UA presentational hints, but I believe it's generally a bad pattern to expose this vocabulary to the language itself.

If we want to consider exposing top layer to random elements, I think that's a discussion that needs to happen independently, with a proper discussion around naming, before exposing it to the language.

**Comment by chrishtr** [comment]
> Also, I'd like to re-iterate my objection to exposing the name "top layer" explicitly to the web platform.

I think another name would be fine. Any ideas? I'll think on it also...

> I think this approach can potentially be revisited

@nt1m I'm not sure which approach you're suggesting could be revisited. Could you clarify?

**Comment by nt1m** [comment]
> I'm not sure which approach you're suggesting could be revisited. Could you clarify?

The one where we append the element in the top layer when the transition starts, and keeping it in the top layer as long as the transition is still running. I originally objected the `:top-layer` pseudo-class interacting with that bit (since it would be misleading), but now that it's `:open`/`:closed` there's a bit more flexibility around this.

**Comment by nt1m** [comment]
Also, looking at the demo in the first comment, it seems like the issue isn't the top layer itself, but rather the individual adjustments the top layer makes. E.g. this CSS works perfectly:

```

@keyframes show {
  0% { 
    opacity: 0; 
  }
}

dialog[open] {
  animation: show 400ms;
  visibility: visible;
  /* UA-default opacity: 1; */
}

@keyframes close {
  0% {
    visibility: visible;
  }
  100% {
    /* Necessary to repeat display: block unless
       interpolation behavior is visibility-like
       preferring non-none value. */
    visibility: visible;
    opacity: 0; }
}

dialog {
  --duration: 400ms;
  animation: close 400ms;
  visibility: hidden;

  /* The UA applies the following while dialog is :modal,
     we need to preserve it during the animation. */
  display: block;
  position: fixed;
  inset-block-start: 0px;
  inset-block-end: 0px;
  max-width: calc((100% - 6px) - 2em);
  max-height: calc((100% - 6px) - 2em);
}
```

so it seems to me this is more about allowing the `display` CSS property to animate?

Also, if this is about having to repeat all the properties that the top layer sets, then don't we have the same issue with any HTML feature that sets UA styling? It seems to me than this is more a problem tied to `display` than top layer.

**Comment by khushalsagar** [comment]
@nt1m it's not just the UA styles applied in top layer, which developers can replicate. It's also the change in stacking which allows the dialog to paint on top of all other content.

**Comment by jods4** [comment]
Is this new take on top-layer animation even less friendly than the previous popover one when it comes to non-time based animations?

In previous spec, it was hacky but possible to control the animation from JS by using "proxy" native animations that one could pause and complete.
If I understand correctly, this new proposal is founded on a CSS `transition` and that, is impossible to control from JS.

Can someone write a basic example of a physics-based animations (e.g. springs) for a top-layer element removal?

**Comment by nt1m** [comment]
Very disappointing feedback here was not taking in account before sending an "Intent to Ship": https://groups.google.com/a/chromium.org/g/blink-dev/c/nx2P-B8Rhx4/m/QbE3n68fAgAJ

**Comment by chrishtr** [comment]
> Very disappointing feedback here was not taking in account before sending an "Intent to Ship": https://groups.google.com/a/chromium.org/g/blink-dev/c/nx2P-B8Rhx4/m/QbE3n68fAgAJ

It is not an intent to ship, it's an intent to prototype. We are building a complete prototype in order to make sure we understand all of the implementation touch points of top layer before coming back to the CSSWG for further discussion.

**Comment by nt1m** [comment]
> > Very disappointing feedback here was not taking in account before sending an "Intent to Ship": [groups.google.com/a/chromium.org/g/blink-dev/c/nx2P-B8Rhx4/m/QbE3n68fAgAJ](https://groups.google.com/a/chromium.org/g/blink-dev/c/nx2P-B8Rhx4/m/QbE3n68fAgAJ)
> 
> It is not an intent to ship, it's an intent to prototype. We are building a complete prototype in order to make sure we understand all of the implementation touch points of top layer before coming back to the CSSWG for further discussion.

Ah right, sorry, must have gotten confused since I saw this coming from https://twitter.com/intenttoship .

**Comment by chrishtr** [comment]
> Ah right, sorry, must have been confused since I saw this coming from https://twitter.com/intenttoship .

No worries. I agree the intenttoship twitter name is a bit confusing.

**Comment by flackr** [comment]
> In previous spec, it was hacky but possible to control the animation from JS by using "proxy" native animations that one could pause and complete. If I understand correctly, this new proposal is founded on a CSS `transition` and that, is impossible to control from JS.

The [getAnimations() API](https://www.w3.org/TR/web-animations-1/#dom-animatable-getanimations) returns all running CSS animations, CSS transitions and web animations allowing them to be modified or controlled from JS. This would let you pause and complete the transition in order to run a JS-based animation.

> Can someone write a basic example of a physics-based animations (e.g. springs) for a top-layer element removal?

Something like this (uses a transition of visibility so that it works in regular browsers but with this proposal the same would work for the top layer transition):
https://output.jsbin.com/qoyevow

**Comment by jods4** [comment]
@flackr thanks!

Good thinking, grabbing the CSS transition with `getAnimations`. 💡
The code is clunky, but at least it works (any non-keyframe-based animations with native web animations is clunky, so nothing unusual there -- I'm hopeful the platform will make some progress on this front in the future).

Your proof of concept works, but for somewhat more "real" code I'd tweak it a bit:
Listening to `transitionstart` is a good idea but then you should just look for transitions that target `visibility` (which you know is the key piece here) rather than using `evt.propertyName`. In a real application, the dialog might well start other transitions for other states so that would not be reliable.

**Comment by flackr** [comment]
Just to put the sample code in the issue here, it would look something like this for top-layer:
```js
dialog.addEventListener('transitionstart', (evt) => {
  if (evt.propertyName != 'top-layer')
    return;
  let transition = evt.target.getAnimations().filter((anim) => anim.transitionProperty == evt.propertyName)[0];
  transition.pause();
  runCustomPhysicsJSAnimation().then(() => { transition.finish(); });
});
```

You could also roughly (extra work to ignore already running animations not included here) polyfill the previous way this worked by using `getAnimations({subtree: true})` to find all animations within the dialog and adjusting the transition duration to match the end time of the last animation, e.g.
```js
dialog.addEventListener('transitionstart', (evt) => {
  if (evt.propertyName != 'top-layer')
    return;
  let transition = evt.target.getAnimations().filter((anim) => anim.transitionProperty == evt.propertyName)[0];
  let maxDuration = evt.target.getAnimations({subtree: true}).map(anim => anim.effect.getComputedTiming().endTime).reduce((maxDuration, duration) => Math.max(duration, maxDuration), 0);
  transition.effect.updateTiming({duration: maxDuration});
});
```

**Comment by lilles** [comment]
Here is what I wrote up about spec changes for CSS and HTML that would support transitions of top layer elements with a top-layer property. It matches the implementation we have behind a flag in Blink.

**TL;DR** Outline spec changes needed to introduce the top layer concept in CSS with a `top-layer` property as well as specifying how ordering of top layer candidate elements can be provided by the host languages like HTML.

## Background

* The order of elements in the top layer is currently given by the host language via the HTML and Fullscreen specifications.
* Currently, those specifications also say which elements are actually rendered in the top layer or not.
* An element is currently still considered to be in the top layer in a given order even when not rendered. A dialog can e.g. be in the top layer even when it is deep down in a `display: none` subtree.
* Moving elements in and out of the top layer happens synchronously per current specifications and is not connected to the update rendering step in the HTML specification.
* There is a proposal to be able to transition elements out of the top layer and have the element still be rendered in the top layer during the transition.
* The `top-layer` property proposal is in [issue #8189](https://github.com/w3c/csswg-drafts/issues/8189)
* There is a [long-standing request in the fullscreen specification](https://fullscreen.spec.whatwg.org/#rendering) to define the top layer concept in CSS as part of the stacking context model.

## Proposed Changes

### CSS Specification

* Introduce the concept of rendering elements in the top layer, how such elements interact with the box tree, stacking, positioning, etc. The current [Fullscreen specification](https://fullscreen.spec.whatwg.org/#new-stacking-layer) can be used as a starting point.
* Introduce a `top-layer` property taking the values `none` and `browser`.
* Require that the host language provides an ordered list of elements from the DOM, along with the DOM, which are candidates for being rendered in the CSS top layer.
* If the computed value of the `top-layer` property for an element in that ordered list is `browser`, render the element in the top layer. If the element is in that list but the `top-layer` property computes to `none`, do normal rendering as part of the DOM tree.
* It is up to the host language to specify UA styles.
* It is up to the host language to move elements in and out of the ordered list of candidates for top layer rendering.


### HTML Specification

* Add two collections of top layer elements
    1. The ordered list of top layer candidate elements to be passed to the CSS implementation.
    2. A set of candidates to be removed from list (i) after the rendering has been updated. This is a subset of the elements in (i).
* When the current HTML specification says to add an element to the top layer:
    1. If the element is in the top layer candidate elements list, and not in the set of candidate elements to be removed, do nothing.
    2. Otherwise add it to the end of the candidate element list, but remove it from the candidate list first if already present.
* When the current HTML specification says to remove an element to the top layer, add the element to the set of candidates to be removed instead.
* After the update rendering step, remove all elements in the set of candidates to be removed for which either apply (remove both from the to-be-removed set and the ordered candidate list):
    1. The element is not rendered
    2. The computed value of `top-layer` is `none`
* Introduce UA styles as outlined in [issue #8189](https://github.com/w3c/csswg-drafts/issues/8189) to not allow author to generally override top layer rendering while allowing the author to keep elements in the top layer for exit transitions by adding `transition-property: top-layer`


### Fullscreen Specification

* Interacts with the HTML specification. Possibly some details about immediately removing elements from the top layer candidate list if top layer transitions should not work for fullscreen'ed elements.

**Comment by nt1m** [comment]
Again, can we come up with a different name than top layer?

**Comment by chrishtr** [comment]
How about `overlay`?

`overlay: top`: in the top layer
`overlay: normal`: in normal z-index flow

Fullscreen, `popover` and `dialog` can all be thought of as UI overlays.

**Comment by nt1m** [comment]
`overlay` is a lot better, will think more on my side and come back with more suggestions if I find any.

**Comment by nt1m** [comment]
I think `overlay: auto` and `overlay: none` would be more accurate than `top`/`normal` since `top` won't work on all elements (but only on top layer ones).

**Comment by nt1m** [comment]
Potentially `overlay-behavior: auto` makes more sense than `overlay: auto`.

This new property is essentially about disabling top layer layout/rendering behaviors during transitions, so I think including `-behavior` in there makes sense.

**Comment by nt1m** [comment]
Fwiw, I think top layer is a very bad name here because this property really is about the rendering behaviors that top layer has, rather than the top layer elements list itself. You can have an element in the top layer list, while having those top layer rendering behaviors turned off during transitions using that property.

So `overlay-behavior: auto/none` seems to express this better IMO.

**Comment by nt1m** [comment]
@chrishtr I'm trying the `top-layer` property in Chrome Canary, and I'm not sure how it works exactly, but my example doesn't seem to work: https://jsfiddle.net/n1p67m0L/

**Comment by lilles** [comment]
> @chrishtr I'm trying the `top-layer` property in Chrome Canary, and I'm not sure how it works exactly, but my example doesn't seem to work: https://jsfiddle.net/n1p67m0L/

- Discrete transitions are not in Canary yet (https://chromium-review.googlesource.com/c/chromium/src/+/4219315)
- display:none animations are behind a test flag (can be enabled with --enable-blink-features=CSSDisplayAnimation), not an experimental flag in Canary.
- In order for your demo to work with the WIP CL above applied, and CSSDisplayAnimation feature enabled, I also had to add a transition for the display property since the UA style for the dialog has display:none when it's closed.
- The dialog is repositioned during the opacity transition because positioning also changes in the UA style when the dialog is closed.

**Comment by chrishtr** [comment]
A complete working demo is [here](https://codepen.io/chrishtr1/pen/eYLyGgq) that you can use on Chrome Canary. Command-line to test:

```
chrome --user-data-dir=/tmp/foo --enable-blink-features=CSSDisplayAnimation,CSSTransitionDiscrete,CSSTopLayerForTransitions https://codepen.io/chrishtr1/pen/eYLyGgq
```

This demonstrates a dialog that is normally below a green overlay, but when it is in the top layer it is on top. It also demonstrates animation of discrete properties generally, `display` (to prevent disappearing), and `top-layer` (to stay on top of the green overlay).

When closing the dialog, it transition-animates opacity, display and top-layer with potentially different durations for each.

**Comment by chrishtr** [comment]
I propose that we move forward with a new `overlay-behavior` CSS property specified according to the rules described [here](https://github.com/w3c/csswg-drafts/issues/8189#issuecomment-1447989290) (except with change of property name of course) and with syntax like this:

`overlay-behavior: auto | none`

* Auto would mean "keep in the top layer except when force-ejected by the UA". None is a no-op from regular behavior.
* The default is "none".
* UAs can force-remove an element in situations like an animation taking too long to complete.

There is the following UA rule:

```
* {
  overlay-behavior: none !important;
}
```

Note: the only way to override such a UA rule is with `transition: overlay-behavior`.

**Comment by nt1m** [comment]
@chrishtr Just wondering, instead of creating a new property, has syncing with `display` been considered? Mainly putting this out there since it has been done for inert: https://github.com/w3c/csswg-drafts/issues/8389

**Comment by chrishtr** [comment]
> @chrishtr Just wondering, instead of creating a new property, has syncing with `display` been considered? Mainly putting this out there since it has been done for inert: #8389

Interesting idea. I think it would work in cases where top layer elements animate to `display:none`, but not cases where they don't. e.g. a developer could add `dialog { display:block }` or `[popover] { display:block }` to their style sheet. In that situation, it's unclear when to remove the element from the top layer when dismissing the dialog or popover, without the UA having to try to guess or detect when animations are happening on dismiss.

**Comment by fantasai** [comment]
> Note: the only way to override such a UA rule is with transition: overlay-behavior.

To be clear, that can't override the UA rule. You can't override a UA !important rule, only the UA can override it, so if the UA doesn't apply an `overlay-behavior: auto !important` rule to an element it will stay as `none`. But the author can control the transition between `none` and `auto` with such a rule.

From an initial read, I think the proposal makes sense. I think using the term “overlay” is a great improvement over “top-layer”. :) I'm a little uncertain about -behavior as the suffix, it doesn't seem to be behavioral, more like a new positioning scheme. (I wonder if `overlay-index: none | auto` makes sense? It's reminiscent of z-index, except the UA is picking the index according to the host rules. Idk)

**Comment by nt1m** [comment]
> > @chrishtr Just wondering, instead of creating a new property, has syncing with `display` been considered? Mainly putting this out there since it has been done for inert: #8389
> 
> Interesting idea. I think it would work in cases where top layer elements animate to `display:none`, but not cases where they don't. e.g. a developer could add `dialog { display:block }` or `[popover] { display:block }` to their style sheet. In that situation, it's unclear when to remove the element from the top layer when dismissing the dialog or popover, without the UA having to try to guess or detect when animations are happening on dismiss.

If your element has `display: block` even before the transition, you're going to see a visible shift either way because of all the different adjustments top layer makes: position, containing-block, stacking context, etc. Adding a new property can control the timing of that shift, but cannot remove it. So I don't think this is an use case we should worry too much about.

**Comment by chrishtr** [comment]
> If your element has `display: block` even before the transition, you're going to see a visible shift either way because of all the different adjustments top layer makes: position, containing-block, stacking context, etc. Adding a new property can control the timing of that shift, but cannot remove it. So I don't think this is an use case we should worry too much about.

In some cases, maybe, but I think preventing the z-index shift during animation will still be useful to developers in these cases. And there's still the problem I mentioned earlier of not being able to easily distinguish between an animation to `display:none` vs animation to something else.

**Comment by LeaVerou** [comment]
Btw do note that `*` does not target every possible selector target. E.g. what happens if I do `::before { overlay: auto }` ? Is the plan to add every possible pseudo-element in the UA stylesheet so it can get `overlay: initial !important`?

**Comment by nt1m** [comment]
Instead of a new property (which I'm still not a fan of, because it's non-intuitive for web developers), I'm thinking:
- sync z-order adjustment of top layer to `z-index` property
- sync other adjustments to `display` property (like done for inert)

I think this is what probably feels the most natural, and it should be somewhat simple to do this in WebKit. @chrishtr What do you think?

**Comment by chrishtr** [comment]
Hi @nt1m,

Can you tell me more about how z-order syncing and adjustments to `display` would work? Are you suggesting something that automatically detects animations from developers and syncs to them?

**Comment by nt1m** [comment]
> Hi @nt1m,
> 
> Can you tell me more about how z-order syncing and adjustments to `display` would work? Are you suggesting something that automatically detects animations from developers and syncs to them?

Have all the non-z-order ones sync with display (display: block = adjust, display: none = do not adjust) in a similar fashion inert is synced with display. Have the z-order ones sync magically with the z-index property (this would probably involve some magic internal value).

**Comment by chrishtr** [comment]
> Have all the non-z-order ones sync with display (display: block = adjust, display: none = do not adjust) in a similar fashion inert is synced with display. Have the z-order ones sync magically with the z-index property (this would probably involve some magic internal value).

I see. Sounds like the same suggestion you made earlier in the issue? I think my responses [here](https://github.com/w3c/csswg-drafts/issues/8189#issuecomment-1464607425) and [here](https://github.com/w3c/csswg-drafts/issues/8189#issuecomment-1466313439) are reasons why I think this is not as clean and simple as the `overlay` CSS property in semantics, "magic" or reliability for web developers.

I also don't agree that `overlay` will be hard for developers to understand, they just need to use the `transition` CSS property combined with it, which is straightforward and easy for them to copy-and-paste.

**Comment by nt1m** [comment]
> > Have all the non-z-order ones sync with display (display: block = adjust, display: none = do not adjust) in a similar fashion inert is synced with display. Have the z-order ones sync magically with the z-index property (this would probably involve some magic internal value).
> 
> I see. Sounds like the same suggestion you made earlier in the issue?

Not really, you mentioned developers want to control the `z-index` shift separately from `display` so my suggestion is to control this separately from the `display` property.

> I also don't agree that `overlay` will be hard for developers to understand, they just need to use the `transition` CSS property combined with it, which is straightforward and easy for them to copy-and-paste.

I'm not concerned about it being hard to understand. I'm trying to fit top layer into the pre-existing the CSS models, which I do think it leads to a better developer experience it we manage to do so, rather than introducing a new model. E.g. imagine being a web developer knowing nothing about the top layer (which is expected in most cases), but knowing basic CSS, what would be their first intuition to solving the problem?

I think it would be good to think about this first before moving forward with a new property, I'm open to any suggestions that moves towards that direction.

**Comment by flackr** [comment]
Something like `z-index: overlay` which would be a `z-index` value only available for the UA to apply? Then developers could add a transition on that which would preserve the overlay setting on exiting.

**Comment by nt1m** [comment]
> Something like z-index: overlay which would be a z-index value only available for the UA to apply? Then developers could add a transition on that which would preserve the overlay setting on exiting.
> Downside would be that naively other z-index changes would also transition.

I don't see that as a downside. I think this makes sense given z-index doesn't apply in the top layer, and top layer overrides any other z-index.

What made me hesitant of suggesting this was more how top layer does more than changing the z-index. But thinking more about this, it's not uncommon in CSS to have properties that cause side-effects.

**Comment by nt1m** [comment]
> A new specialized property, on the other hand, avoids this concern - you can naively set a transition on the property, and it only applies in the expected situation.

I do see the issue, but toying around with z-index outside of the top layer and toggling with that same element in & out of the top layer seems like an unlikely situation. The most common case is having a static z-index or none and toggling in & out of the top layer.

> All told z-index is similar thematically to top layer, but in virtually every specific it's different, and conflating them is imo likely to lead to significantly more confusion than not.

I'm aware of the specifics (having implemented top layer in WebKit) and also of the style adjustments associated with top layer, though not sure it's significant from a web developer perspective.

> A new property requires new learning, sure, but the purpose of the new property is very narrow in the first place (it exists solely to allow an author to add a transition that delays its value change, which is pretty narrow and specific).

As you mentioned in your first edit, it is weird :) It is also unprecedented in CSS to have properties that exist only for transitions. This is why it would be nice to think a bit more about it. I'm not advocating specifically for the `z-index` proposal, but more for an alternative proposal. A new property is _ok_ but the standard should be higher for a spec, given they last once you implement them on the web.

**Comment by tabatkins** [comment]
> I'm aware of the specifics

Right, not suggesting you weren't, just saying that this is better considered a completely novel functionality.

> A new property is ok but the standard should be higher for a spec, given they last once you implement them on the web.

The alternative to "a property that exists only for transitions" (and which can have this condition enforced using existing mechanisms like the UA-!important style sheet) is "a property value that exists only for transitions" (and which has to have novel magic defined to enforce this condition). I'm not even sure what the full degree of the magic *is*, actually; we want to prevent authors from setting it (so the top-layer remains managed by the UA) which could possibly be a constraint on the computed value, but we also want to prevent authors from *un*setting it (for the same reason, so we can keep the top layer consistent when, for example, there are stacks that should close together), and I don't know how I'd do that. (I suspect the most straightforward way would be to say elements have a hidden state bit that, when turned on, forces z-index to compute to 'overlay' regardless of what the author does, and it's prevented from computing to 'overlay' otherwise? But that state bit functions identically to the 'overlay' property here.)

Both of a property and a value last once they're implemented, to the exact same degree. I don't think learnability is significantly different between the two? But being able to lean on existing functionality rather than forcing some *really* novel cascading behavior into one specific value seems like a big benefit of using a new property.

**Comment by chrishtr** [comment]
> Both of a property and a value last once they're implemented, to the exact same degree. I don't think learnability is significantly different between the two? But being able to lean on existing functionality rather than forcing some _really_ novel cascading behavior into one specific value seems like a big benefit of using a new property.

Just wanted to +1 this point. The thing I like the most about the `overlay` property approach is that it doesn't require any new implementation or spec machinery to do with animations, and just re-uses what is already there. Likewise, it re-uses a concept of transitions that developers using animations are already familiar with. And the only new stuff for browsers to implement is how it affects what's in the top layer during the transition, which is I think a browser implementation and spec complexity that any proposal would need to have.

**Comment by astearns** [comment]
@nt1m I am removing the agenda+ label for now, as the discussion here seems to be going fine and I do not see consensus on an alternative to what we resolved before.

**Comment by tabatkins** [comment]
All the top-layer definitions, along with the 'overlay' property, are now in the Position 4 spec: <https://drafts.csswg.org/css-position-4/#top-layer>

There have been several minor changes from the text as it existed in Fullscreen.

* the 'overlay' property now exists, which is automatically set by the UA (and *only* the UA) to reflect whether the element is in the top layer or not, but which can be affected by author-level transitions to keep something in the top layer temporarily, while an exit animation is playing (fading its opacity, etc). 'overlay' has transition semantics similar to 'visibility', where all intermediate progress values map to 'auto' (rather than flipping at half-progress)
	* as part of this, the "pending top layer removals" list has been added, so other algorithms can tell whether something is *actually* still in the top layer or just pending removal due to a transition; if it's pending they can go ahead and act as if it wasn't there.
* modifying the top layer has been abstracted behind a couple of algorithms. They're generally light-touch, with two major changes:
	* unless you explicitly ask for immediate removal, removal instead just adds it to the pending removal list, and it's removed for real during HTML's Update The Rendering. This ensures that transitions added at the same time will get processed and can have an effect on the 'overlay' property, possibly delaying the *actual* removal. (Immediate removal should only be used for security-related things, or when CSS isn't relevant, like if the element has been removed from the document.)
	* adding something to the top layer that's already in the top layer (and not pending removal) is now considered a spec error, enforced by an assertion. (Previously, it just shifted the element to the end of the top layer, so it rendered topmost.) Now that several features play in the top layer, their interactions are non-trivial, and should be handled more explicitly by each feature. This will require more spec work, probably mostly in the other features' specs.

**Comment by chrishtr** [comment]
I'm going to close this issue now since the spec is defined and accepted by the WG. Further discussion can continue in issue #8730 if needed.
